pkg/cloud/baremetal/actuators/machine/actuator.go

```diff
@@ -584,3 +593,163 @@ func (a *Actuator) nodeAddresses(host *bmh.BareMetalHost) ([]corev1.NodeAddress,
 
 	return addrs, nil
 }
+
+//deleteRemediationAnnotations deletes remediation-in-progress and remediation strategy annotations
+func (a *Actuator) deleteRemediationAnnotations(ctx context.Context, machine *machinev1.Machine) error {
+	if len(machine.Annotations) == 0 {
+		return nil
+	}
+
+	delete(machine.Annotations, remediationInProgressAnnotation)
+	delete(machine.Annotations, externalRemediationAnnotation)
+
+	if err := a.client.Update(ctx, machine); err != nil {
+		log.Printf("Failed to delete annotations of Machine: %s", machine.Name)
+		return err
+	}
+
+	return nil
+}
+
+//hasRebootAnnotation checks if the reboot annotation exist on the baremetalhost
+func hasRebootAnnotation(baremetalhost *bmh.BareMetalHost) (exists bool) {
+	if len(baremetalhost.Annotations) > 0 {
+		_, exists = baremetalhost.Annotations[rebootAnnotation]
+	}
+	return
+}
+
+//addRemediationInProgressAnnotation adds a remediation-in-progress annotation to the machine
+func (a *Actuator) addRemediationInProgressAnnotation(ctx context.Context, machine *machinev1.Machine) error {
+	if machine.Annotations == nil {
+		machine.Annotations = make(map[string]string)
+	}
+
+	machine.Annotations[remediationInProgressAnnotation] = ""
+
+	err := a.client.Update(ctx, machine)
+	if err != nil {
+		log.Printf("Failed to add remediation in progess annotation to %s: %s", machine.Name, err.Error())
+	}
+
+	return err
+}
+
+//requestPowerOn removes reboot annotation on baremetalhost which signal BMO to power on the machine
+func (a *Actuator) requestPowerOff(ctx context.Context, baremetalhost *bmh.BareMetalHost) error {
+	if baremetalhost.Annotations == nil {
+		baremetalhost.Annotations = make(map[string]string)
+	}
+
+	baremetalhost.Annotations[rebootAnnotation] = ""
+
+	err := a.client.Update(ctx, baremetalhost)
+	if err != nil {
+		log.Printf("failed to add reboot annotation to %s: %s", baremetalhost.Name, err.Error())
+	}
+
+	return err
+}
+
+//requestPowerOn adds reboot annotation on baremetalhost which signal BMO to power off the machine
+func (a *Actuator) requestPowerOn(ctx context.Context, baremetalhost *bmh.BareMetalHost) error {
+	if baremetalhost.Annotations == nil {
+		baremetalhost.Annotations = make(map[string]string)
+	}
+
+	if _, rebootPending := baremetalhost.Annotations[rebootAnnotation]; !rebootPending {
+		return &clustererror.RequeueAfterError{}
+	}
+
+	delete(baremetalhost.Annotations, rebootAnnotation)
+
+	err := a.client.Update(ctx, baremetalhost)
+	if err != nil {
+		log.Printf("failed to remove reboot annotation from %s: %s", baremetalhost.Name, err.Error())
+	}
+
+	return err
+}
+
+// deleteMachineNode deletes the node that mapped to specified machine
+func (a *Actuator) deleteNode(ctx context.Context, node *corev1.Node) error {
+	err := a.client.Delete(ctx, node)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return &clustererror.RequeueAfterError{}
```

Comment:
If the node was already deleted, we know that we can advance to the next remediation stage (requesting power on), so we want a requeue to continue the flow.

Reply:
But if the Node was deleted successfully, you want to return `nil`? What's different about that case?

Reply:
Good point. I'll need to change it to requeue as well

Reply:
Just occurred to me: not sure how long it might take to delete the Node, but it would be a good idea to check its deletion timestamp and just requeue after a small delay in case it's already set.

Reply:
If the `Delete()` function returned no error, doesn't it mean that deletion timestamp was set?
What value can we gain from checking the deletion timestamp vs just assume it was set after `Delete()`?

Reply:
The next time reconciliation runs, the Node resource might still exist with its deletion timestamp set. In that case, rather than calling Delete again, you could just requeue with a small delay. It saves an extra delete call at least, and probably would prevent this code from spinning in a hot loop, calling Delete and re-queueing over and over as fast as possible until deletion completes.

Reply:
Nice catch. Fixed